[BUG] Fix CAGRA search recall with a graph built by NN Descent

enp1s0 opened the pull request:
This PR addresses an unexpected low recall issue in the CAGRA search with a graph generated by NN Descent.
For that, it updates the initial NN Descent graph generation so that all indices are included, whereas some nodes are not found in the branch-25.10 implementation.


<img width="691" height="316" alt="figure" src="https://github.com/user-attachments/assets/0b9f2d6b-9ace-4c28-8deb-a6b5b0241bc6" />

(itopk=256)

This PR also makes the following changes:
- Adds a workaround to remove duplicate node indices from the kNN result
- Adds test cases for different graph degrees

enp1s0 committed:
Fix `init_random_graph`

enp1s0 committed:
Fix NND init

enp1s0 committed:
Update NND init

enp1s0 committed:
Remove duplicates

enp1s0 committed:
Add degree check in NND

enp1s0 committed:
Add setting degree in CAGRA test

tfeher commented on cpp/src/neighbors/detail/nn_descent.cuh:
```diff
@@ -829,36 +830,29 @@ void GnndGraph<Index_t>::sample_graph_new(InternalID_t<Index_t>* new_neighbors,
 template <typename Index_t>
 void GnndGraph<Index_t>::init_random_graph()
```
Could you add a docstring to describe what properties the random graph is expected to have? What is ensured by the current PR to fix the low recall issue? E.g. is it guaranteed to be fully connected?

tfeher reviewed (changes requested):
Thank you @enp1s0 for this PR! Looks good overall, I have just one request to clarify the expectations for the random graph.

KyleFromNVIDIA reviewed (approved):
All CMake/CI/packaging changes have been removed, so approving.

enp1s0 committed:
Fix CI test

enp1s0 committed:
Fix init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update the explanation of init

enp1s0 committed:
Change random graph init method

enp1s0 committed:
Fix assert

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update init

enp1s0 committed:
Fix dtype

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Fix copyright

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Remove unnecessary assert

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update test recall threshold

enp1s0 committed:
Improve complete graph build

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

AyodeAwe committed:
Update Changelog [skip ci]

Intron7 committed:
Add KDE kernel (#1915)

@cjnolet moved the kernel for KDE from cuml to cuvs. This PR is needed for https://github.com/rapidsai/cuml/pull/7833

Authors:
  - Severin Dicks (https://github.com/Intron7)
  - Anupam (https://github.com/aamijar)

Approvers:
  - Victor Lafargue (https://github.com/viclafargue)
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuvs/pull/1915

jnke2016 committed:
Align docs with pluggable benchmark API (#1891)

**Description of changes**

cuVS Bench documentation updates to align with the pluggable benchmark API:

- **index.rst**: Fix datasets.yaml path and document the `constraints` field in algorithm YAML; add Recommended (orchestrator API) and Legacy CLI workflow; fix intro typo.
- **param_tuning.rst**: Add Benchmark modes (sweep vs tune) and tune-mode example.
- **pluggable_backend.rst** (new): Document config loader + backend pair, run flow, what the loader produces, steps to add a backend, and a minimal Elasticsearch-style example. Add to cuvs_bench toctree.

Authors:
  - Joseph Nke (https://github.com/jnke2016)
  - Corey J. Nolet (https://github.com/cjnolet)
  - Anupam (https://github.com/aamijar)

Approvers:
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuvs/pull/1891

benfred committed:
Coderabbit integration (#1908)

Following on the work of https://github.com/rapidsai/rmm/pull/2244 and https://github.com/rapidsai/cuml/pull/7725, this enables general instructions and configuration to enable coderabbit codereviews on cuvs.

Closes https://github.com/rapidsai/cuvs/issues/1767

Authors:
  - Ben Frederickson (https://github.com/benfred)
  - Corey J. Nolet (https://github.com/cjnolet)
  - Anupam (https://github.com/aamijar)

Approvers:
  - Anupam (https://github.com/aamijar)
  - Corey J. Nolet (https://github.com/cjnolet)
  - Bradley Dice (https://github.com/bdice)

URL: https://github.com/rapidsai/cuvs/pull/1908

cjnolet commented:
@enp1s0 there still seems to be a C++ test failure in the int8 CAGRA test. Can you take a look?

jinsolp committed:
Add docs for cagra mem usage with NN Descent build algo (#2000)

Adding cagra mem usage for building with the NN Descent algorithm.
Improving formatting of existing docs.


Previous:
<img width="824" height="565" alt="Screenshot 2026-04-08 at 4 04 55 PM" src="https://github.com/user-attachments/assets/0d18de28-e5c6-4ee5-8f90-8896c6e4c059" />


Updated in the PR:

<img width="807" height="565" alt="Screenshot 2026-04-08 at 4 05 18 PM" src="https://github.com/user-attachments/assets/1f3d0eaa-4c90-4a1d-ac18-32f3df0a612c" />

Authors:
  - Jinsol Park (https://github.com/jinsolp)

Approvers:
  - Anupam (https://github.com/aamijar)
  - Corey J. Nolet (https://github.com/cjnolet)

URL: https://github.com/rapidsai/cuvs/pull/2000

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update CAGRA test

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Fix potential overflow

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update NND for return_distance=1

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Update json files

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

achirkin committed:
Merge branch 'main' into cagra-fix-nnd-init

achirkin commented on cpp/src/neighbors/detail/cagra/jit_lto_kernels/compute_distance_matrix.json:
```diff
@@ -142,6 +144,12 @@
         "pq_prefix": "_vpq",
         "pq_suffix": "_8pq_4subd"
+      },
+      {
+        "pq_len": "8",
+        "pq_bits": "8",
+        "pq_prefix": "_vpq",
+        "pq_suffix": "_8pq_8subd"
       }
```
I guess these changes leaked from the same updates in https://github.com/rapidsai/cuvs/pull/1533 accidentally?

achirkin reviewed (approved):
LGTM and sorry you've had to keep the PR in sync with changes in cuvs for so long!

One note: could you please check one more time that all changes to the instantiation matrices are intentional?

enp1s0 committed:
Revert "Update json files"

This reverts commit b4d348c71213a7255784ac47c9fcb93602018d14.

enp1s0 committed:
Merge branch 'cagra-fix-nnd-init' of github.com:enp1s0/cuvs into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

tfeher reviewed (approved):
LGTM, thanks @enp1s0 for the updates.

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

aamijar committed:
Merge branch 'main' into cagra-fix-nnd-init

aamijar committed:
fix copyright

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

aamijar committed:
Merge branch 'main' into cagra-fix-nnd-init

aamijar committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'cagra-fix-nnd-init' of github.com:enp1s0/cuvs into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'main' into cagra-fix-nnd-init

achirkin committed:
Merge branch 'release/26.08' into cagra-fix-nnd-init

achirkin committed:
Merge branch 'release/26.08' into cagra-fix-nnd-init

enp1s0 committed:
Merge branch 'release/26.08' into cagra-fix-nnd-init

achirkin committed:
Bound index deduplication attempts to avoid infinite loops on small datasets (n_rows <= graph_degree)
